### PR TITLE
Reduce ByteBuffer duplication when resizing pooled direct ByteBufs

### PR DESCRIPTION
**Pull request by njhill**
Motivation

Currently when use of Unsafe is disabled and an internal reallocation is performed for a direct `PooledByteBuf`, a one-off temporary duplicate is made of the source and destination backing nio buffers so that the copy can be done in a threadsafe manner.

The need for this can be reduced by sharing the temporary duplicate buffer that is already stored in the corresponding destination `PooledByteBuf` instance.

Modifications

Have `PoolArena#memoryCopy(...)` take the destination `PooledByteBuf` instead of the underlying mem reference and offset, and use `internalNioBuffer()` to obtain/initialize a reusable duplicate of the
backing nio buffer.

Result

Fewer temporary allocations when resizing direct pooled `ByteBuf`s in the non-`Unsafe` case

### CODE REVIEW

**Comment by normanmaurer** [comment]
@netty-bot test this please

File: buffer/src/main/java/io/netty/buffer/PoolArena.java [hunk 1]
```diff
@@ -402,7 +402,9 @@ void reallocate(PooledByteBuf<T> buf, int newCapacity, boolean freeOldMemory) {
             buf.trimIndicesToCapacity(newCapacity);
             bytesToCopy = newCapacity;
         }
-        memoryCopy(oldMemory, oldOffset, buf.memory, buf.offset, bytesToCopy);
+        if (bytesToCopy > 0) {
+            oldNioBuffer = memoryCopy(oldMemory, oldNioBuffer, oldOffset, buf, bytesToCopy);
```
Comment (normanmaurer) [inline buffer/src/main/java/io/netty/buffer/PoolArena.java]: hmm is this really correct ? This may now be duplicate that is passed to free(...)

Comment (njhill) [reply buffer/src/main/java/io/netty/buffer/PoolArena.java]: @normanmaurer the `oldNioBuffer` passed there is already a duplicate (comes from `buf.tmpNioBuf`) and isn't free'd itself, rather kept in the cache for later re-use by other `ByteBuf`s that share the same "root" nio buffer...

**Comment by normanmaurer** [comment]
@njhill honestly I wonder if this change really worth it... I mean it makes things more complex and resizing should be "hopefully" not happen too often. Or I am missing anything ?

**Comment by njhill** [comment]
@normanmaurer that's a fair point... I noticed this while looking closer at the underlying buffer manipulation done in `ByteToMessageDecoder` cumulation but you are probably right re the relative frequency not being big high to matter. Let me remove the "more complex" half of the change ... the other part is a slight simplification if anything imho

**Comment by normanmaurer** [comment]
@netty-bot test this please

**Comment by normanmaurer** [comment]
@njhill can you update the commit message and pr description to match what you did now ?

**Comment by njhill** [comment]
@normanmaurer done

**Comment by normanmaurer** [comment]
@netty-bot test this please

**Comment by normanmaurer** [comment]
@njhill thanks.. and sorry for the delay.

**Pull request merged**